File: LC/[1473] Paint House III.cpp
class Solution {
    /*
    DP: min(cost) of
        + m houses
        + target blocks
        + n colors
    */
    int dp[105][105][25];//100 house, target neighbors, n colors
public:
    int minCost(vector<int>& houses, vector<vector<int>>& cost, int m, int n, int target) 
    {
        houses.insert(houses.begin(), 0); //whenever you need to ref to prev row
        cost.insert(cost.begin(), {0});

        for (int i=0; i<=m; i++) //initialize DP
            for (int j=0; j<=target; j++)
                for (int k=0; k<=n; k++)
                    dp[i][j][k] = INT_MAX/2;//min cost so initialize inf

        if (houses[1]!=0) //houses[0] is dummy
        {
            dp[1][1][houses[1]] = 0;//already painted (all others are invalid)
        }
        else
        {
            for (int k=1; k<=n; k++)//not painted, populate all colors
                dp[1][1][k] = cost[1][k-1];
        }

        //for each house
        for (int i=2; i<=m; i++)//remember ii=0 is dummy, ii=1 is pre-processed
        {
            if (houses[i]!=0)//current house is painted
            {       
                for (int j=1; j<=target; j++)
                {
                    int k = houses[i];                    
                    for (int kk=1; kk<=n; kk++) //prev house color
                    {
                        if (kk==k)//if current color same as prev house color (same block)
                            dp[i][j][k] = min(dp[i][j][k], dp[i-1][j][kk]);
                        else//current color not same as prev color
                            dp[i][j][k] = min(dp[i][j][k], dp[i-1][j-1][kk]); //(diff block)
                    }
                }
            }
            else//current house need to be painted
            {
                //for each block
                for (int j=1; j<=target; j++)
                    for (int k=1; k<=n; k++)//for each color
                    {
                        for (int kk=1; kk<=n; kk++)//for each previous color
                        {
                            if (kk==k)//if prev color same as current color
                                dp[i][j][k] = min(dp[i][j][k], dp[i-1][j][kk] + cost[i][k-1]);//same block
                            else
                                dp[i][j][k] = min(dp[i][j][k], dp[i-1][j-1][kk] + cost[i][k-1]);//diff block
                        }
                    }                
            }    
        }

        int ret = INT_MAX/2;
        for (int k=1; k<=n; k++)//for each color
            ret = min(ret, dp[m][target][k]);//last house, last block, each color
        if (ret==INT_MAX/2)
            return -1;
        else
            return ret;
    }
};

File: LC/[1473] Paint House III.py
class Solution:
    def minCost(self, houses: List[int], cost: List[List[int]], m: int, n: int, target: int) -> int:
        """
        # I/O:
        + M houses
        + N colors
        ++ painted or unpainted
        + Target Blocks
        
        # Analysis:
        + current state only dependes on previous state -> Type I DP

        """
        # insert dummy
        houses=[math.inf]+houses   #for prev house
        cost = [[math.inf]] + cost #cost also start from 1 as 0 is dummy

        m=len(houses)

        # declare dp
        dp=[[[math.inf]*(n+1) for _ in range(target+1)] for _ in range(m)]
        
        # initialize DP 
        if houses[1] != 0: #first house is colored? (remember 0 is dummy)
            dp[1][1][houses[1]] = 0 # house 1 already painted, cost=0, all other color on house1 are invalid
        else:
            for kk in range(1, n+1):
                dp[1][1][kk]=cost[1][kk-1]
        

        # populate DP
        for ii in range(2, m): # for each house, 0 is dummy, 1 is initialized
            if houses[ii] != 0: #house is painted
                k = houses[ii] #current color
                for jj in range(1, target+1): #for each target                  
                    for kk in range(1, n+1): # previous color
                        if k==kk: #same color as prev house, same block
                            dp[ii][jj][k] = min(dp[ii][jj][k], dp[ii-1][jj][kk]) # no new cost added to same block
                        else: #diff color as prev house, new block
                            dp[ii][jj][k] = min(dp[ii][jj][k], dp[ii-1][jj-1][kk]) #no cost added to new block
            else: # house is NOT painted
                for jj in range(1, target+1):   # for each block
                    for k in range(1, n+1):    # for each prev color
                        for kk in range(1, n+1): # for each curr color
                            if k==kk: #curr color same as prev color
                                dp[ii][jj][k] = min(dp[ii][jj][k], dp[ii-1][jj][kk] + cost[ii][k-1])
                            else: # curr color diff than prev color
                                dp[ii][jj][k] = min(dp[ii][jj][k], dp[ii-1][jj-1][kk] + cost[ii][k-1])
        
        ans = math.inf

        for kk in range(1, n+1):
            ans = min(ans, dp[-1][-1][kk])
        
        #print(dp)

        return ans if ans!=math.inf else -1
